src/test/msixtest/api_packagewriter.cpp

```diff
@@ -0,0 +1,352 @@
+//
+//  Copyright (C) 2019 Microsoft.  All rights reserved.
+//  See LICENSE file in the project root for full license information.
+// 
+//  Validates IAppxPackageWriter interface
+#include "catch.hpp"
+#include "msixtest_int.hpp"
+#include "FileHelpers.hpp"
+#include "PackTestData.hpp"
+#include "macros.hpp"
+
+#include <iostream>
+
+using namespace MsixTest::Pack;
+
+constexpr std::uint32_t DefaultBlockSize = 65535;
+
+void InitializePackageWriter(IStream* outputStream, IAppxPackageWriter** packageWriter)
+{
+    *packageWriter = nullptr;
+
+    MsixTest::ComPtr<IAppxFactory> appxFactory;
+    REQUIRE_SUCCEEDED(CoCreateAppxFactoryWithHeap(MsixTest::Allocators::Allocate, MsixTest::Allocators::Free,
+        MSIX_VALIDATION_OPTION_SKIPSIGNATURE, &appxFactory));
+    REQUIRE_SUCCEEDED(appxFactory->CreatePackageWriter(outputStream, nullptr, packageWriter));
+    return;
+}
+
+// Test creating instances of package writer
+TEST_CASE("Api_AppxPackageWriter_create", "[api]")
+{
+    MsixTest::ComPtr<IAppxFactory> appxFactory;
+    REQUIRE_SUCCEEDED(CoCreateAppxFactoryWithHeap(MsixTest::Allocators::Allocate, MsixTest::Allocators::Free,
+        MSIX_VALIDATION_OPTION_SKIPSIGNATURE, &appxFactory));
+
+    auto outputStream = MsixTest::StreamFile("test_package.msix", false, true);
+    MsixTest::ComPtr<IAppxPackageWriter> packageWriter;
+
+    // Test incorrect nullptr parameter. Note, currently we don't support APPX_PACKAGE_SETTINGS in the MSIX SDK
+    // we always create a zip64 package and use SHA256 as hash method, so nullptr as second parameter is fine.
+    REQUIRE_HR(static_cast<HRESULT>(MSIX::Error::InvalidParameter),
+        appxFactory->CreatePackageWriter(nullptr, nullptr, &packageWriter));
+    REQUIRE_HR(static_cast<HRESULT>(MSIX::Error::InvalidParameter),
+        appxFactory->CreatePackageWriter(outputStream.Get(), nullptr, nullptr));
+
+    REQUIRE_SUCCEEDED(appxFactory->CreatePackageWriter(outputStream.Get(), nullptr, &packageWriter));
+}
+
+// Test creating a valid msix package via IAppxPackageWriter with different file sizes
+TEST_CASE("Api_AppxPackageWriter_good", "[api]")
+{
+    auto outputStream = MsixTest::StreamFile("test_package.msix", false, true);
+
+    MsixTest::ComPtr<IAppxPackageWriter> packageWriter;
+    InitializePackageWriter(outputStream.Get(), &packageWriter);
+
+    // These values are set so that the files added to the package have increasingly
+    // larger sizes, with the first file having a small size < DefaultBlockSize, and
+    // the last file having a large size > 10x DefaultBlockSize.
+    const std::uint32_t contentSizeIncrement = DefaultBlockSize * 10 / static_cast<uint32_t>(TestConstants::GoodFileNames.size()) + 1;

```

Comment:
>const std::uint32_t contentSizeIncrement = DefaultBlockSize * 10 / static_cast<uint32_t>(TestConstants::GoodFileNames.size()) + 1; [](start = 4, length = 130)

This is a strangely arbitrary value that can change over time. Why?

Reply:
Following the windows tests. I would guess that it was to guarantee that no matter how many filenames are added the last one will be10xDefaultBlockSize-ish

---
In reply to: [302264428](https://github.com/Microsoft/msix-packaging/pull/170#discussion_r302264428) [](ancestors = 302264428)